File: sdk/node/framework/session.ts
/**
 * AgentSession - Session wrapper with DI support
 *
 * Wraps the core session with DI container and scope management.
 */

import { DIContainer } from './container';

export class AgentSession {
  private guards: any[] = [];
  private interceptors: any[] = [];
  private pipes: any[] = [];
  private filters: any[] = [];

  constructor(
    private sessionId: string,
    private workspace: string,
    private container: DIContainer,
    private scopeId: string
  ) {}

  /**
   * Run the session with a prompt
   */
  async run(prompt: string): Promise<any> {
    // TODO: Execute middleware pipeline
    // TODO: Call Rust core session.run()

    // For now, return a mock result
    return {
      text: `Mock response for: ${prompt}`,
      sessionId: this.sessionId,
      workspace: this.workspace,
    };
  }

  /**
   * Add a guard to this session
   */
  useGuard<T>(guardClass: new (...args: any[]) => T): this {
    const guardInstance = this.container.resolve(guardClass, this.scopeId);
    this.guards.push(guardInstance);
    return this;
  }

  /**
   * Add an interceptor to this session
   */
  useInterceptor<T>(interceptorClass: new (...args: any[]) => T): this {
    const interceptorInstance = this.container.resolve(interceptorClass, this.scopeId);
    this.interceptors.push(interceptorInstance);
    return this;
  }

  /**
   * Add a pipe to this session
   */
  usePipe<T>(pipeClass: new (...args: any[]) => T): this {
    const pipeInstance = this.container.resolve(pipeClass, this.scopeId);
    this.pipes.push(pipeInstance);
    return this;
  }

  /**
   * Add an exception filter to this session
   */
  useFilter<T>(filterClass: new (...args: any[]) => T): this {
    const filterInstance = this.container.resolve(filterClass, this.scopeId);
    this.filters.push(filterInstance);
    return this;
  }
}

File: sdk/python/a3s_code_framework/session.py
"""
AgentSession - Express-like session wrapper

Wraps the core session with middleware pipeline support.
"""

from typing import Any, Callable, Optional


class AgentSession:
    """Agent Session

    Express-like session wrapper with middleware pipeline.
    """

    def __init__(
        self,
        session_id: str,
        workspace: str,
    ):
        """Initialize session

        Args:
            session_id: Session ID
            workspace: Workspace directory path
        """
        self._session_id = session_id
        self._workspace = workspace

        # TODO: Initialize Rust core session when bindings are ready
        self._core_session = None

        # Middleware pipeline
        self._middleware = []

    def use(self, middleware: Callable) -> "AgentSession":
        """Add middleware to the pipeline (Express-style)

        Args:
            middleware: Middleware function or instance

        Returns:
            Self for chaining
        """
        self._middleware.append(middleware)
        return self

    async def run(self, prompt: str) -> Any:
        """Run the session with a prompt

        Args:
            prompt: User prompt

        Returns:
            Session result
        """
        # TODO: Execute middleware pipeline
        # TODO: Call Rust core session.run()

        # For now, return a mock result
        return {
            "text": f"Mock response for: {prompt}",
            "session_id": self._session_id,
            "workspace": self._workspace,
        }
